File: backend/app/schemas/queue.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class QueueEntryCreate(BaseModel):
    patient_id: UUID
    doctor_id: UUID | None = None
    appointment_id: UUID | None = None
    notes: str | None = Field(default=None, max_length=1000)


class QueueStatusUpdate(BaseModel):
    status: str  # in_consultation | completed | cancelled


class QueueEntryResponse(BaseModel):
    id: UUID
    clinic_id: UUID
    branch_id: UUID | None = None
    patient_id: UUID
    patient_name: str | None = None
    doctor_id: UUID | None = None
    doctor_name: str | None = None
    appointment_id: UUID | None = None
    queue_number: int
    status: str
    notes: str | None = None
    called_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class QueueListResponse(BaseModel):
    data: list[QueueEntryResponse]
    total: int


class QueuePositionResponse(BaseModel):
    """Patient-facing live queue position (GET /queue/my-position).

    All fields are null/empty (ahead_count=0) when the patient has no queue
    entry today — the endpoint returns 200, not 404, in that case."""

    queue_entry_id: UUID | None
    clinic_id: UUID | None
    clinic_name: str | None
    doctor_name: str | None
    queue_number: int | None
    position: int | None
    status: str | None
    ahead_count: int
    estimated_wait_minutes: int | None


class QueueDisplayToken(BaseModel):
    """One anonymized token on the waiting-room board.

    Token labels are derived from the daily queue_number only — no patient
    identifiers, names, or notes ever leave this payload (PHI-safe by
    construction, suitable for a public TV)."""

    token: str  # e.g. "Q-12"
    status: str  # waiting | in_consultation
    position: int | None = None  # 1-based place among waiting entries
    called_at: datetime | None = None


class QueueDisplayResponse(BaseModel):
    """Clinic waiting-room board payload (GET /queue/display)."""

    now_serving: list[QueueDisplayToken]
    up_next: list[QueueDisplayToken]
    waiting_count: int  # total waiting (up_next may be truncated by `limit`)
    generated_at: datetime

File: frontend/src/lib/api/queue.ts
import api from "@/lib/api";

export interface QueueEntry {
  id: string;
  clinic_id: string;
  branch_id: string | null;
  patient_id: string;
  patient_name: string | null;
  doctor_id: string | null;
  doctor_name: string | null;
  appointment_id: string | null;
  queue_number: number;
  status: "waiting" | "in_consultation" | "completed" | "cancelled";
  notes: string | null;
  called_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface QueueListResponse {
  data: QueueEntry[];
  total: number;
}

/**
 * Patient-facing live queue position (GET /api/v1/queue/my-position).
 * All fields are null/empty when the patient isn't checked in anywhere today.
 */
export interface MyQueuePosition {
  queue_entry_id: string | null;
  clinic_id: string | null;
  clinic_name: string | null;
  doctor_name: string | null;
  queue_number: number | null;
  position: number | null;
  status: QueueEntry["status"] | null;
  ahead_count: number;
  estimated_wait_minutes: number | null;
}

export async function getMyQueuePosition(): Promise<MyQueuePosition> {
  const res = await api.get("/api/v1/queue/my-position");
  return res.data;
}

/**
 * Anonymized waiting-room board (GET /api/v1/queue/display).
 * Token labels only — the payload carries no patient identifiers, names,
 * or notes, so it is safe to render on a public-facing TV.
 */
export interface QueueDisplayToken {
  token: string; // e.g. "Q-12"
  status: "waiting" | "in_consultation";
  position: number | null; // 1-based place among waiting entries
  called_at: string | null;
}

export interface QueueDisplayResponse {
  now_serving: QueueDisplayToken[];
  up_next: QueueDisplayToken[];
  waiting_count: number; // total waiting (up_next may be truncated by limit)
  generated_at: string;
}

export async function getQueueDisplay(limit = 8): Promise<QueueDisplayResponse> {
  // X-Clinic-Id is attached automatically by the axios interceptor.
  const res = await api.get("/api/v1/queue/display", { params: { limit } });
  return res.data;
}

export async function getQueue(
  clinicId: string,
  params?: { status?: string; doctor_id?: string }
): Promise<QueueListResponse> {
  const res = await api.get("/api/v1/queue", {
    headers: { "X-Clinic-Id": clinicId },
    params,
  });
  return res.data;
}

export async function addToQueue(
  clinicId: string,
  data: {
    patient_id: string;
    doctor_id?: string;
    appointment_id?: string;
    notes?: string;
  }
): Promise<QueueEntry> {
  const res = await api.post("/api/v1/queue", data, {
    headers: { "X-Clinic-Id": clinicId },
  });
  return res.data;
}

export async function updateQueueStatus(
  clinicId: string,
  entryId: string,
  status: "in_consultation" | "completed" | "cancelled"
): Promise<QueueEntry> {
  const res = await api.patch(
    `/api/v1/queue/${entryId}/status`,
    { status },
    { headers: { "X-Clinic-Id": clinicId } }
  );
  return res.data;
}

export async function removeFromQueue(
  clinicId: string,
  entryId: string
): Promise<void> {
  await api.delete(`/api/v1/queue/${entryId}`, {
    headers: { "X-Clinic-Id": clinicId },
  });
}
